docs/guides/basic-rbac.mdx

```diff
@@ -3,267 +3,303 @@ title: Implement basic Role Based Access Control (RBAC) with metadata
 
-![The Users page in the Clerk Dashboard with the 'Edit public metadata' modal open. The modal has a text field with the JSON 'role': 'admin'.](/docs/images/guides/basic-rbac/edit-public-metadata.webp)
+  ```ts {{ filename: 'utils/roles.ts' }}
+  import { Roles } from '@/types/global'
+  import { auth } from '@clerk/nextjs/server'
 
-## Create an admin dashboard and protect it
+  export const checkRole = (role: Roles) => {
+    const { sessionClaims } = auth()
+    return sessionClaims?.metadata.role === role
+  }
+  ```
 
-Now that your user has the `admin` role, let's build an admin dashboard to help you easily change users' roles. In your `app/` directory, create an `admin/` folder. Within the `admin/` folder, create a `dashboard/` folder. In the `dashboard/` folder, create a file named `page.tsx`. Copy the following code and paste it into the file.
+  > [!NOTE]
+  > You can customize the behavior of the `checkRole()` helper function to suit your needs. For example, you could modify it to return the roles a user has or create a `protectByRole()` function that handles role-based redirects.
 
-You want the dashboard to only be available to users with the `admin` role. [As you configured earlier](#configure-the-session-token), a user's role can be found in the metadata stored in the session token. You can access the session token's claims using Clerk's `auth()` hook.
+  ### Create the admin dashboard
 
-```tsx {{ filename: 'app/admin/dashboard/page.tsx' }}
-import { auth } from '@clerk/nextjs/server'
-import { redirect } from 'next/navigation'
+  Now, it's time to create an admin dashboard. The first step is to create the `/admin` route.
 
-export default function AdminDashboard() {
-  const { sessionClaims } = auth()
+  1. In your `app/` directory, create an `admin/` folder.
+  1. In the `admin/` folder, create a file named `page.tsx`.
+  1. Add the following placeholder code to the file.
 
-  // If the user does not have the admin role, redirect them to the home page
-  if (sessionClaims?.metadata.role !== 'admin') {
-    redirect('/')
+  ```tsx {{ filename: 'app/admin/page.tsx' }}
+  export default function AdminDashboard() {
+    return <p>This is the protected admin dashboard restricted to users with the `admin` role.</p>
   }
+  ```
+
+  ### Protect the admin dashboard
+
+  To protect the `/admin` route, choose **one** of the two following methods:
 
-  return (
-    <>
-      <h1>This is the admin dashboard</h1>
-      <p>This page is restricted to users with the 'admin' role.</p>
-    </>
-  )
-}
-```
+  1. **Middleware**: Apply role-based access control globally at the route level. This method restricts access to all routes matching `/admin` before the request reaches the actual page.
+  1. **Page-Level Role Check**: Apply role-based access control directly in the admin page component. This method protects the page itself but doesn't block access at the middleware level.
```

Comment:
```suggestion
  1. **Page-Level Role Check**: Apply role-based access control directly in the admin page component. This method protects the page itself but does. not protect other pages/routes in the admin dashboard. If you also have a `/admin/invoices` route you would need to also apply this protection to that route.
```
That was a quick edit. The bit about Middleware is correct, but I think pointing out the need to protect other pages is more helpful here. That said, reword as you feel best.

Reply:
[b20141c](https://github.com/clerk/clerk-docs/pull/1560/commits/b20141cd7ae85be7bc56f4a6feb71805fbee63c6)